examples/raytrace.dx

```diff
@@ -180,143 +127,231 @@ def sampleCosineWeightedHemisphere (k:Key) (normal: Vec3) : Vec3 =
+    HitLight intensity -> intensity
+    HitNothing -> zero
+    HitObj _ _ -> zero
+
+def sampleSquare (hw:Real) (k:Key) : Position =
+ (kx, kz) = splitKey k
+ x = randuniform (- hw) hw kx
+ z = randuniform (- hw) hw kz
+ [x, 0.0, z]
+
+def sampleLightRadiance
+      (scene:Scene n) (osurf:OrientedSurface) (inRay:Ray) (k:Key) : Radiance =
+  (surfNor, surf) = osurf
+  (rayPos, _) = inRay
+  (MkScene objs) = scene
+  snd $ withAccum \radiance.
+    for i. case objs.i of
+      PassiveObject _ _ -> ()
+      Light lightPos hw _ ->
+        (dirToLight, distToLight) = directionAndLength $
+                                      lightPos + sampleSquare hw k - rayPos
+        case positiveProjection dirToLight surfNor of
+          False -> () -- light on the far side of current surface
+          True ->
+            fracSolidAngle = (relu $ dot dirToLight yHat) * sq hw / (pi * sq distToLight)
+            outRay = (rayPos, dirToLight)
+            coeff = fracSolidAngle * probReflection osurf inRay outRay
+            radiance += coeff .* rayDirectRadiance scene outRay
+
+def trace (params:Params) (scene:Scene n) (init_ray:Ray) (k:Key) : Color =
+  (_, max_bounces) = params
+  -- TODO: we ought to be able to use an accumulator here, but there's a bug
+  noFilter = [1.0, 1.0, 1.0]
+  iter (noFilter, zero, init_ray) $
+    \i (filter, radiance, ray).
+      case i >= max_bounces of
+        True -> Done radiance
+        False -> case raymarch scene ray of
+          HitNothing -> Done radiance
+          HitLight intensity -> case i == 0 of
+            True  -> Done intensity  -- TODO: scale etc
+            False -> Done radiance
+          HitObj incidentRay osurf ->
+            (k1, k2) = splitKey $ hash k i
+            lightRadiance    = sampleLightRadiance scene osurf incidentRay k1
+            outRayHemisphere = sampleReflection          osurf incidentRay k2
+            newFilter = surfaceFilter filter (snd osurf)
+            newRadiance = radiance + applyFilter newFilter lightRadiance
+            Continue (newFilter, newRadiance, outRayHemisphere)
+
+def avgRayColor (params:Params) (scene:Scene m) (ray:Ray) (k:Key) : Color =
+  (num_samples, _) = params
+  sampleAveraged (trace params scene ray) num_samples k
+
+-- TODO: add number of pixels once we can hide sizes
+-- sensor half-width, pinhole-sensor distance, pinhole position
+-- (Assumes we're looking towards -z.)
+Camera = (Position & Real & Real)
+
+def cameraRays (n:Int) (camera:Camera) : Fin n => Fin n => Ray =
+  -- images indexed from top-left
+  (pos, halfWidth, sensorDist) = camera
+  ys = reverse $ linspace (Fin n) (neg halfWidth) halfWidth
+  xs =           linspace (Fin n) (neg halfWidth) halfWidth
+  for i j. (pos, normalize [xs.j, ys.i, neg sensorDist])
+
+def takePicture
+       (params:Params) (scene:Scene m) (n:Int) (camera:Camera)
+      : ColorImage n n =
+  rays = cameraRays n camera
+  rootKey = newKey 0
+  for i j.
+    k = ixkey (ixkey rootKey i) j
```

Comment:
What if you used the same key for every pixel?  I think that would still be unbiased, but would result in crisp images even when only using a few samples.

Reply:
Good idea! Very reparameterization-trick-ish. Here's the rendering with the same number of pixels/samples/bounces as before but with the same key for each pixel. It introduces different artifacts, of course, but I think they're quite pleasing to look at.

![rendering2](https://user-images.githubusercontent.com/8833761/89362382-d3ce2980-d69b-11ea-8275-1460e8ca0106.png)

Reply:
This should be called "the David trick" because he always suggests it and it always works.

Reply:
I got the general heuristic to derandomize wherever possible from E.T. Jaynes:  "It appears to be a quite general principle that, whenever there is a randomized way of doing something, then there is a nonrandomized way that delivers better performance but requires more thought."

Reply:
@bachili just pointed me to a ray tracer using a similar trick: https://users.cg.tuwien.ac.at/zsolnai/gfx/smallpaint/ . They get a similar "painted" look:

> The cause of the painterly look is actually a bug – there’s an incorrect usage of the Halton-series with correlating dimensions, and the errors are spread throughout the screen with an equally wrong sampling function. one would wish that every bug could look just as beautiful as this one.